### PR TITLE
ICU-21683 ICU-21674 Clear state from UFormattedNumber[Range] in C API

### PR DESCRIPTION
**Pull request by sffc**
##### Checklist

- [x] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-21683, https://unicode-org.atlassian.net/browse/ICU-21674
- [x] Required: The PR title must be prefixed with a JIRA Issue number. 
- [x] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [x] Required: Each commit message must be prefixed with a JIRA Issue number. 
- [x] Issue accepted (done by Technical Committee after discussion)
- [x] Tests included, if applicable
- [ ] API docs and/or User Guide docs changed or added, if applicable

### CODE REVIEW

File: icu4c/source/i18n/formattedval_sbimpl.cpp [hunk 1]
```diff
@@ -283,14 +288,14 @@ bool FormattedValueStringBuilderImpl::nextPositionImpl(ConstrainedFieldPosition&
 
 void FormattedValueStringBuilderImpl::appendSpanInfo(UFieldCategory category, int32_t spanValue, int32_t start, int32_t length, UErrorCode& status) {
     if (U_FAILURE(status)) { return; }
-    U_ASSERT(spanIndices.getCapacity() >= spanValue);
-    if (spanIndices.getCapacity() == spanValue) {
-        if (!spanIndices.resize(spanValue * 2, spanValue)) {
+    U_ASSERT(spanIndices.getCapacity() >= spanIndicesCount);
+    if (spanIndices.getCapacity() == spanIndicesCount) {
+        if (!spanIndices.resize(spanIndicesCount * 2, spanIndicesCount)) {
```
Comment (FrankYFTang) [inline icu4c/source/i18n/formattedval_sbimpl.cpp:293]: Do you also need to changre the spanValue to spanIndicesCount in the following three lines inside the next function prependSpanInfo ?

```
    U_ASSERT(spanIndices.getCapacity() >= spanValue);
    if (spanIndices.getCapacity() == spanValue) {
        if (!spanIndices.resize(spanValue * 2, spanValue)) {
```

File: icu4c/source/i18n/formattedval_sbimpl.cpp [hunk 2]
```diff
@@ -96,6 +96,11 @@ void FormattedValueStringBuilderImpl::getAllFieldPositions(FieldPositionIterator
     }
 }
 
+void FormattedValueStringBuilderImpl::reset() {
+    fString.clear();
+    spanIndicesCount = 0;
```
Comment (FrankYFTang) [inline icu4c/source/i18n/formattedval_sbimpl.cpp:101]: does it make sense to call it "reset"? If it's functionality is truely to "reset" FormattedValueStringBuilderImpl then should it also clear out the  fNumericField and spanIndices too? If fNumericField and spanIndices should not be clear out, then maybe the name "reset" does not truly reflect what you like it to do.

Comment (FrankYFTang) [reply icu4c/source/i18n/formattedval_sbimpl.cpp:101]: Also I see there are subclass adding more fields to the object, so if it is named "reset" I would expect the function also reset the extra fields the subclass added. Maybe call it resetXXX to "limit" what the scope it intend to reset of it's member data?

**Comment by sffc** [comment]
@markusicu can you take this one?  @FrankYFTang was in the middle of the review but is now OOO.

**Review by markusicu — APPROVED** [review]
lgtm

**Pull request merged**